<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Marnin Wolfe" />

<meta name="date" content="2021-08-13" />

<title>Empirically-driven simulation of an existing cassava GS program</title>

<script src="site_libs/header-attrs-2.10/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    background-color: #232629;
    color: #7a7c7d;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #7a7c7d;  padding-left: 4px; }
div.sourceCode
  { color: #cfcfc2; background-color: #232629; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #cfcfc2; } /* Normal */
code span.al { color: #95da4c; background-color: #4d1f24; font-weight: bold; } /* Alert */
code span.an { color: #3f8058; } /* Annotation */
code span.at { color: #2980b9; } /* Attribute */
code span.bn { color: #f67400; } /* BaseN */
code span.bu { color: #7f8c8d; } /* BuiltIn */
code span.cf { color: #fdbc4b; font-weight: bold; } /* ControlFlow */
code span.ch { color: #3daee9; } /* Char */
code span.cn { color: #27aeae; font-weight: bold; } /* Constant */
code span.co { color: #7a7c7d; } /* Comment */
code span.cv { color: #7f8c8d; } /* CommentVar */
code span.do { color: #a43340; } /* Documentation */
code span.dt { color: #2980b9; } /* DataType */
code span.dv { color: #f67400; } /* DecVal */
code span.er { color: #da4453; text-decoration: underline; } /* Error */
code span.ex { color: #0099ff; font-weight: bold; } /* Extension */
code span.fl { color: #f67400; } /* Float */
code span.fu { color: #8e44ad; } /* Function */
code span.im { color: #27ae60; } /* Import */
code span.in { color: #c45b00; } /* Information */
code span.kw { color: #cfcfc2; font-weight: bold; } /* Keyword */
code span.op { color: #cfcfc2; } /* Operator */
code span.ot { color: #27ae60; } /* Other */
code span.pp { color: #27ae60; } /* Preprocessor */
code span.re { color: #2980b9; background-color: #153042; } /* RegionMarker */
code span.sc { color: #3daee9; } /* SpecialChar */
code span.ss { color: #da4453; } /* SpecialString */
code span.st { color: #f44f4f; } /* String */
code span.va { color: #27aeae; } /* Variable */
code span.vs { color: #da4453; } /* VerbatimString */
code span.wa { color: #da4453; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">BreedingSchemeOpt</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="AlphaSimHlpR_firstSteps.html">AlphaSimHlpR first steps</a>
    </li>
    <li>
      <a href="example-simulation-reducing-error-with-new-tools.html">Reducing error with new tools</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Simulations
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="baselineSim.html">Baseline simulation</a>
    </li>
  </ul>
</li>
<li>
  <a href="resources.html">Resources</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Other
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="about.html">About</a>
    </li>
    <li>
      <a href="license.html">License</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/wolfemd/BreedingSchemeOpt">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Empirically-driven simulation of an existing cassava GS program</h1>
<h4 class="author">Marnin Wolfe</h4>
<h4 class="date">2021-08-13</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-08-19
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>BreedingSchemeOpt/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20210422code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20210422)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20210422code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20210422)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomwolfemdBreedingSchemeOpttree5914d8df72caa55375fb4d7336f1f8f5755e8763targetblank5914d8da"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/wolfemd/BreedingSchemeOpt/tree/5914d8df72caa55375fb4d7336f1f8f5755e8763" target="_blank">5914d8d</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomwolfemdBreedingSchemeOpttree5914d8df72caa55375fb4d7336f1f8f5755e8763targetblank5914d8da" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/wolfemd/BreedingSchemeOpt/tree/5914d8df72caa55375fb4d7336f1f8f5755e8763" target="_blank">5914d8d</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/images/.DS_Store
    Ignored:    data/.DS_Store

Untracked files:
    Untracked:  data/baselineScheme.gsheet
    Untracked:  output/benchmark_sim.rds
    Untracked:  output/benchmark_sims5.rds

Unstaged changes:
    Modified:   data/README.md

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/baselineSim.Rmd</code>) and HTML (<code>docs/baselineSim.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/BreedingSchemeOpt/blob/5914d8df72caa55375fb4d7336f1f8f5755e8763/analysis/baselineSim.Rmd" target="_blank">5914d8d</a>
</td>
<td>
wolfemd
</td>
<td>
2021-08-19
</td>
<td>
Publish initial sims towards a baseline set of sims using runBreedingScheme_wBurnIn
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/BreedingSchemeOpt/blob/c2e379e19de5a854ca2cc037d4072c66704b0e01/analysis/baselineSim.Rmd" target="_blank">c2e379e</a>
</td>
<td>
wolfemd
</td>
<td>
2021-08-13
</td>
<td>
Rebuild repo removing the “Group” part. Project is to contain BOTH the “Group” and a separate section just for the actual simulation analyses.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>See <a href="https://wolfemd.github.io/IITA_2021GS/inputsForSimulation.html">here</a> for an analysis of IITA trial data to derive empirical inputs for this analysis. I analyzed trial-by-trial to measure selection error and associate it with plot-size. While inconclusive, this exercise emphasized a key concern for conducting simulations that alter the VDP: that the cost-benefit balance could depend on the relative information value/selection accuracy/error variance of different plot sizes and trial configurations.</p>
<p>As a result, I decided to simulate a range of error-vs-plot size scaling as part of the “baseline” simulations, starting with IITA as an example.</p>
<p><strong><em>If</em></strong> we observe a shift-point in the cost-benefit analysis we can then work with breeding programs to determine where their data indicate they lie and what changes are subsequently recommended.</p>
<div id="develop-a-simulation-with-burn-in-ps" class="section level1">
<h1>Develop a simulation with burn-in PS</h1>
<ul>
<li>First thing is to complete un-finished work (<a href="example-simulation-reducing-error-with-new-tools.html">started here</a>) and build a <a href="https://wolfemd.github.io/AlphaSimHlpR/reference/runBreedingScheme_wBurnIn.html"><code>runBreedingScheme_wBurnIn()</code></a> function into <code>AlphaSimHlpR</code>. The new function enables a switch in selection criteria after a certain number of cycles, e.g. a phenotypic selection “burn-in” period followed by GS.</li>
<li>Previously, used control files to set-up <code>bsp</code>. Implemented <a href="https://wolfemd.github.io/AlphaSimHlpR/reference/specifyBSP.html"><code>specifyBSP()</code></a> which creates a <code>bsp</code> using a <code>data.frame</code> of stage-specific breeding scheme <em>plus</em> all other <code>AlphaSimHlpR</code> arguments as inputs.</li>
<li>I fully documented all new functions, integrated them into my forked-repo of <code>AlphaSimHlpR</code> <em>and</em> built a <code>pkgdown</code> web-documentation: <a href="https://wolfemd.github.io/AlphaSimHlpR/index.html">here</a></li>
<li>(click the function refs above to see their specific details)</li>
<li>Below, a quick demo before subsequently setting up a bigger analysis</li>
</ul>
<div id="set-up-a-singularity-shell-with-ropenblas" class="section level2">
<h2>Set-up a singularity shell with R+OpenBLAS</h2>
<p>This is not required. If you want the advantage of multi-threaded BLAS to speed up predictions within the simulations, you need an R instance that is linked to OpenBLAS (another example is Microsoft R Open). For CBSU, the recommended approach is currently to use singularity shells provided by the “rocker” project. They even come pre-installed with tidyverse :).</p>
<p>Linked to OpenBLAS, using a simple function <code>RhpcBLASctl::blas_set_num_threads()</code> I can add arguments to functions to control this feature.</p>
<p>For optimal performance, it is import to balance the number of threads each R session uses for BLAS against any other form of parallel processing being used and considering total available system resources.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 0) Pull a singularity image with OpenBLAS enabled R + tidyverse from rocker/</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> pull ~/rocker2.sif docker://rocker/tidyverse:latest<span class="kw">;</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># only do above first time</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 1) start a screen shell </span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">screen</span><span class="kw">;</span> <span class="co"># or screen -r if re-attaching...</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 3) start the singularity Linux shell inside that</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> shell ~/rocker2.sif<span class="kw">;</span> </span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Project directory, so R will use as working dir.</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /home/mw489/BreedingSchemeOpt/<span class="kw">;</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># 3) Start R</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install genomicMateSelectR to user-accessible libPath</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="do">### In a singularity shell, sintall as follows:</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>libPath<span class="ot">&lt;-</span><span class="st">&quot;/home/mw489/R/x86_64-pc-linux-gnu-library/4.1&quot;</span> <span class="co"># should be YOUR libPath</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>withr<span class="sc">::</span><span class="fu">with_libpaths</span>(<span class="at">new=</span>libPath, devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;wolfemd/genomicMateSelectR&quot;</span>, <span class="at">ref =</span> <span class="st">&#39;master&#39;</span>))</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="do">### Else, simply</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;wolfemd/genomicMateSelectR&quot;</span>, <span class="at">ref =</span> <span class="st">&#39;master&#39;</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Install my own forked repo of AlphaSimHlpR</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>withr<span class="sc">::</span><span class="fu">with_libpaths</span>(<span class="at">new=</span>libPath, <span class="fu">install.packages</span>(<span class="st">&quot;Rcpp&quot;</span>))</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>withr<span class="sc">::</span><span class="fu">with_libpaths</span>(<span class="at">new=</span>libPath, <span class="fu">install.packages</span>(<span class="st">&quot;AlphaSimR&quot;</span>))</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>withr<span class="sc">::</span><span class="fu">with_libpaths</span>(<span class="at">new=</span>libPath, <span class="fu">install.packages</span>(<span class="st">&quot;optiSel&quot;</span>))</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>withr<span class="sc">::</span><span class="fu">with_libpaths</span>(<span class="at">new=</span>libPath, <span class="fu">install.packages</span>(<span class="st">&quot;rgl&quot;</span>))</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>withr<span class="sc">::</span><span class="fu">with_libpaths</span>(<span class="at">new=</span>libPath, devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;wolfemd/AlphaSimHlpR&quot;</span>, <span class="at">ref =</span> <span class="st">&#39;master&#39;</span>, <span class="at">force=</span>T))</span></code></pre></div>
</div>
<div id="a-small-example" class="section level2">
<h2>A small example</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(AlphaSimHlpR))</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(tidyverse))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(genomicMateSelectR))</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>select <span class="ot">&lt;-</span> dplyr<span class="sc">::</span>select</span></code></pre></div>
<p>Run two cycles pre- and post with a small breeding scheme. Runs on a laptop.</p>
<p>Use my newly created <code>specifyBSP()</code> function to create the <code>bsp</code> input for sims.</p>
<ul>
<li>3 chrom, Ne = 100, 300 SNP (100/chrom)</li>
<li>Select 10 parents, make 4 random crosses with 50 progeny each</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>schemeDF<span class="ot">&lt;-</span><span class="fu">read.csv</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data&quot;</span>,<span class="st">&quot;baselineScheme - Test.csv&quot;</span>), </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">header =</span> T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>bsp<span class="ot">&lt;-</span><span class="fu">specifyBSP</span>(<span class="at">schemeDF =</span> schemeDF,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">nChr =</span> <span class="dv">3</span>,<span class="at">effPopSize =</span> <span class="dv">100</span>,<span class="at">quickHaplo =</span> F,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">segSites =</span> <span class="dv">400</span>, <span class="at">nQTL =</span> <span class="dv">40</span>, <span class="at">nSNP =</span> <span class="dv">100</span>, <span class="at">genVar =</span> <span class="dv">40</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">gxeVar =</span> <span class="cn">NULL</span>, <span class="at">gxyVar =</span> <span class="dv">15</span>, <span class="at">gxlVar =</span> <span class="dv">10</span>,<span class="at">gxyxlVar =</span> <span class="dv">5</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">meanDD =</span> <span class="fl">0.5</span>,<span class="at">varDD =</span> <span class="fl">0.01</span>,<span class="at">relAA =</span> <span class="fl">0.5</span>,</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">stageToGenotype =</span> <span class="st">&quot;PYT&quot;</span>,</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">nParents =</span> <span class="dv">10</span>, <span class="at">nCrosses =</span> <span class="dv">4</span>, <span class="at">nProgeny =</span> <span class="dv">50</span>,<span class="at">nClonesToNCRP =</span> <span class="dv">3</span>,</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">phenoF1toStage1 =</span> T,<span class="at">errVarPreStage1 =</span> <span class="dv">500</span>,</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">useCurrentPhenoTrain =</span> F, </span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>                <span class="at">nCyclesToKeepRecords =</span> <span class="dv">30</span>,</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">selCritPipeAdv =</span> selCritIID, </span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>                <span class="at">selCritPopImprov =</span>  selCritIID)</span></code></pre></div>
<p>I created a CSV to specify a data.frame <code>schemeDF</code> defining stage-specific breeding scheme inputs.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>schemeDF <span class="sc">%&gt;%</span> rmarkdown<span class="sc">::</span><span class="fu">paged_table</span>()</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["stageNames"],"name":[1],"type":["chr"],"align":["left"]},{"label":["nEntries"],"name":[2],"type":["int"],"align":["right"]},{"label":["nLocs"],"name":[3],"type":["int"],"align":["right"]},{"label":["nReps"],"name":[4],"type":["int"],"align":["right"]},{"label":["nChks"],"name":[5],"type":["int"],"align":["right"]},{"label":["entryToChkRatio"],"name":[6],"type":["int"],"align":["right"]},{"label":["errVars"],"name":[7],"type":["int"],"align":["right"]},{"label":["trainingPopCycles"],"name":[8],"type":["int"],"align":["right"]},{"label":["PlantsPerPlot"],"name":[9],"type":["int"],"align":["right"]}],"data":[{"1":"PYT","2":"200","3":"1","4":"1","5":"2","6":"20","7":"100","8":"15","9":"10"},{"1":"AYT","2":"100","3":"1","4":"1","5":"2","6":"20","7":"50","8":"15","9":"25"},{"1":"UYT","2":"20","3":"1","4":"2","5":"2","6":"20","7":"25","8":"15","9":"42"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Now run <code>runBreedingScheme_wBurnIn()</code> which will simulate PS for burn-in cycles and GS post burn-in. Use <code>selCritIID</code> for VDP. Notice the new <code>parentSelCritGEBV</code> which is so far the same as <code>selCritGRM</code> but lays groundwork for mate selection and non-additive effects related sims.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>test_sim<span class="ot">&lt;-</span><span class="fu">runBreedingScheme_wBurnIn</span>(<span class="at">replication =</span> <span class="dv">1</span>, <span class="at">bsp =</span> bsp,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">nBurnInCycles=</span><span class="dv">2</span>,<span class="at">nPostBurnInCycles=</span><span class="dv">2</span>,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">selCritPopPre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">selCritPopPost=</span><span class="st">&quot;parentSelCritGEBV&quot;</span>,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">selCritPipePre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">selCritPipePost=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">nBLASthreads=</span><span class="dv">3</span>,<span class="at">nThreadsMacs2=</span><span class="dv">3</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(test_sim,<span class="at">file =</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;test_sim.rds&quot;</span>))</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>test_sim<span class="ot">&lt;-</span><span class="fu">readRDS</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;test_sim.rds&quot;</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>test_sim<span class="sc">$</span>records<span class="sc">$</span>stageOutputs <span class="sc">%&gt;%</span> </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(.,<span class="fu">aes</span>(<span class="at">x=</span>year,<span class="at">y=</span>genValMean,<span class="at">color=</span>stage)) <span class="sc">+</span> </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="figure/baselineSim.Rmd/plot%20test%20sim-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This plot shows the mean genetic value (y-axis) by pipeline-stage (colors) versus the year. The vertical line indicates the point after which GS (<code>parentSelCritGEBV</code>) was used.</p>
</div>
</div>
<div id="run-two-cycles-of-iita-sized-gs" class="section level1">
<h1>Run two cycles of IITA-sized GS</h1>
<p>First run a benchmark sim. How long will two cycles of IITA-sized GS rep take?</p>
<p>After benchmarking, I will run a bigger analysis, with multiple iterations times a range of parameter settings. Will use at least one full large memory server for that (112 cores, 512GB RAM). Benchmark using <strong>cbsurobbins.biohpc.cornell.edu</strong>, which now uses SLURM. Reserve and set-up an interactive SLURM shell as follows:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 1) start a screen shell </span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">screen</span><span class="kw">;</span> </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 2) reserve interactive slurm</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="ex">salloc</span> <span class="at">-n</span> 10 <span class="at">--mem</span> 60G<span class="kw">;</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 3) start the singularity Linux shell inside that</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> shell ~/rocker2.sif<span class="kw">;</span> </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Project directory, so R will use as working dir.</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /home/mw489/BreedingSchemeOpt/<span class="kw">;</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 3) Start R</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(AlphaSimHlpR))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(tidyverse))</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(genomicMateSelectR))</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>select <span class="ot">&lt;-</span> dplyr<span class="sc">::</span>select</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># This scheme _excludes_ the seedling stage from the simulation.</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>schemeDF<span class="ot">&lt;-</span><span class="fu">read.csv</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data&quot;</span>,<span class="st">&quot;baselineScheme - IITA.csv&quot;</span>), </span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">header =</span> T, <span class="at">stringsAsFactors =</span> F)</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>schemeDF <span class="sc">%&gt;%</span> rmarkdown<span class="sc">::</span><span class="fu">paged_table</span>()</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["stageNames"],"name":[1],"type":["chr"],"align":["left"]},{"label":["nEntries"],"name":[2],"type":["int"],"align":["right"]},{"label":["nLocs"],"name":[3],"type":["int"],"align":["right"]},{"label":["nReps"],"name":[4],"type":["int"],"align":["right"]},{"label":["nChks"],"name":[5],"type":["int"],"align":["right"]},{"label":["entryToChkRatio"],"name":[6],"type":["int"],"align":["right"]},{"label":["errVars"],"name":[7],"type":["int"],"align":["right"]},{"label":["trainingPopCycles"],"name":[8],"type":["int"],"align":["right"]},{"label":["PlantsPerPlot"],"name":[9],"type":["int"],"align":["right"]}],"data":[{"1":"CET","2":"2500","3":"1","4":"1","5":"5","6":"50","7":"100","8":"15","9":"10"},{"1":"PYT","2":"800","3":"2","4":"3","5":"4","6":"25","7":"100","8":"15","9":"10"},{"1":"AYT","2":"80","3":"3","4":"3","5":"2","6":"20","7":"50","8":"15","9":"25"},{"1":"UYT","2":"30","3":"5","4":"3","5":"2","6":"20","7":"25","8":"15","9":"42"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<ul>
<li>6 chrom, Ne = 1000, 600 SNP (100/chrom)</li>
<li><code>genVar = 100</code> and <code>errVar</code> starting at 100</li>
<li>Select 50 parents, Make 100 random crosses with 25 progeny each</li>
<li><code>nCyclesToKeepRecords = 30</code></li>
<li><code>trainingPopCycles = 15</code> for all stages.</li>
<li>5 cores for BLAS</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>bsp<span class="ot">&lt;-</span><span class="fu">specifyBSP</span>(<span class="at">schemeDF =</span> schemeDF,</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">nChr =</span> <span class="dv">6</span>,<span class="at">effPopSize =</span> <span class="dv">1000</span>,<span class="at">quickHaplo =</span> F,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">segSites =</span> <span class="dv">500</span>, <span class="at">nQTL =</span> <span class="dv">50</span>, <span class="at">nSNP =</span> <span class="dv">100</span>, <span class="at">genVar =</span> <span class="dv">100</span>,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">gxeVar =</span> <span class="cn">NULL</span>, <span class="at">gxyVar =</span> <span class="dv">15</span>, <span class="at">gxlVar =</span> <span class="dv">10</span>,<span class="at">gxyxlVar =</span> <span class="dv">5</span>,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">meanDD =</span> <span class="dv">1</span>,<span class="at">varDD =</span> <span class="dv">5</span>,<span class="at">relAA =</span> <span class="fl">0.1</span>,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">stageToGenotype =</span> <span class="st">&quot;CET&quot;</span>,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">nParents =</span> <span class="dv">50</span>, <span class="at">nCrosses =</span> <span class="dv">100</span>, <span class="at">nProgeny =</span> <span class="dv">25</span>,<span class="at">nClonesToNCRP =</span> <span class="dv">3</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">phenoF1toStage1 =</span> F,<span class="at">errVarPreStage1 =</span> <span class="dv">500</span>,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">useCurrentPhenoTrain =</span> F, </span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">nCyclesToKeepRecords =</span> <span class="dv">30</span>,</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>                <span class="co"># selCrits are overwritten by runBreedingScheme_wBurnIn</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>                <span class="at">selCritPipeAdv =</span> selCritIID, <span class="co"># thus have no actual effect</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">selCritPopImprov =</span>  selCritIID) </span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>starttime<span class="ot">&lt;-</span><span class="fu">proc.time</span>()[<span class="dv">3</span>]</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>benchmark_sim<span class="ot">&lt;-</span><span class="fu">runBreedingScheme_wBurnIn</span>(<span class="at">replication =</span> <span class="dv">1</span>, <span class="at">bsp =</span> bsp,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">nBurnInCycles=</span><span class="dv">2</span>,<span class="at">nPostBurnInCycles=</span><span class="dv">2</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">selCritPopPre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">selCritPopPost=</span><span class="st">&quot;parentSelCritGEBV&quot;</span>,</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">selCritPipePre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">selCritPipePost=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">nBLASthreads=</span><span class="dv">5</span>,<span class="at">nThreadsMacs2=</span><span class="dv">5</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>endtime<span class="ot">&lt;-</span><span class="fu">proc.time</span>()[<span class="dv">3</span>]; <span class="fu">print</span>(<span class="fu">paste0</span>((endtime<span class="sc">-</span>starttime)<span class="sc">/</span><span class="dv">60</span>,<span class="st">&quot; mins elapsed.&quot;</span>));</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(benchmark_sim,<span class="at">file =</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;benchmark_sim.rds&quot;</span>))</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;866.377616666667 mins elapsed.&quot;</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co"># ~14 hrs for 2 cycles of GS.</span></span></code></pre></div>
<p>~14 hrs for 2 cycles of GS. Up to 40 or 50GB RAM.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">readRDS</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;benchmark_sim.rds&quot;</span>))<span class="sc">$</span>records<span class="sc">$</span>stageOutputs <span class="sc">%&gt;%</span> </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(.,<span class="fu">aes</span>(<span class="at">x=</span>year,<span class="at">y=</span>genValMean,<span class="at">color=</span>stage)) <span class="sc">+</span> </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="figure/baselineSim.Rmd/plot%20benchmark%20sim-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This plot shows the mean genetic value (y-axis) by pipeline-stage (colors) versus the year. The vertical line indicates the point after which GS (<code>parentSelCritGEBV</code>) was used. Obviously this is a test. So not digging in more yet.</p>
</div>
<div id="run-a-scaled-down-simulation" class="section level1">
<h1>Run a scaled-down simulation</h1>
<p>It seems prudent to scale down and run additional tests, before scaling back up for full experiments.</p>
<ul>
<li><p>cbsulm09,</p></li>
<li><p>Reduce to <code>trainingPopCycles=5</code></p></li>
<li><p>Reduce pop size to 1/3 scale</p>
<ul>
<li><code>nEntries=ceiling(nEntries/3)</code></li>
<li><code>nChks=ceiling(nChks/3)</code></li>
<li>Select 17 parents, Make 33 random crosses with 26 progeny each</li>
<li>Reduce to 3 chrom, Ne = 500, 300 SNP (100/chrom)</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">screen</span><span class="kw">;</span> </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> shell ~/rocker2.sif<span class="kw">;</span> </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /home/mw489/BreedingSchemeOpt/<span class="kw">;</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(AlphaSimHlpR))</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(tidyverse))</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(genomicMateSelectR))</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>select <span class="ot">&lt;-</span> dplyr<span class="sc">::</span>select</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co"># This scheme _excludes_ the seedling stage from the simulation.</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>schemeDF<span class="ot">&lt;-</span><span class="fu">read.csv</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data&quot;</span>,<span class="st">&quot;baselineScheme - IITA.csv&quot;</span>), </span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">header =</span> T, <span class="at">stringsAsFactors =</span> F) <span class="sc">%&gt;%</span> </span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>PlantsPerPlot) <span class="sc">%&gt;%</span> </span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">trainingPopCycles=</span><span class="dv">5</span>,</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">nEntries=</span><span class="fu">ceiling</span>(nEntries<span class="sc">/</span><span class="dv">3</span>),</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>                <span class="at">nChks=</span><span class="fu">ceiling</span>(nChks<span class="sc">/</span><span class="dv">3</span>))</span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>schemeDF <span class="sc">%&gt;%</span> rmarkdown<span class="sc">::</span><span class="fu">paged_table</span>()</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["stageNames"],"name":[1],"type":["chr"],"align":["left"]},{"label":["nEntries"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["nLocs"],"name":[3],"type":["int"],"align":["right"]},{"label":["nReps"],"name":[4],"type":["int"],"align":["right"]},{"label":["nChks"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["entryToChkRatio"],"name":[6],"type":["int"],"align":["right"]},{"label":["errVars"],"name":[7],"type":["int"],"align":["right"]},{"label":["trainingPopCycles"],"name":[8],"type":["dbl"],"align":["right"]}],"data":[{"1":"CET","2":"834","3":"1","4":"1","5":"2","6":"50","7":"100","8":"5"},{"1":"PYT","2":"267","3":"2","4":"3","5":"2","6":"25","7":"100","8":"5"},{"1":"AYT","2":"27","3":"3","4":"3","5":"1","6":"20","7":"50","8":"5"},{"1":"UYT","2":"10","3":"5","4":"3","5":"1","6":"20","7":"25","8":"5"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>bsp<span class="ot">&lt;-</span><span class="fu">specifyBSP</span>(<span class="at">schemeDF =</span> schemeDF,</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">nChr =</span> <span class="dv">3</span>,<span class="at">effPopSize =</span> <span class="dv">500</span>,<span class="at">quickHaplo =</span> F,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">segSites =</span> <span class="dv">200</span>, <span class="at">nQTL =</span> <span class="dv">30</span>, <span class="at">nSNP =</span> <span class="dv">100</span>, <span class="at">genVar =</span> <span class="dv">100</span>,</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">gxeVar =</span> <span class="cn">NULL</span>, <span class="at">gxyVar =</span> <span class="dv">15</span>, <span class="at">gxlVar =</span> <span class="dv">10</span>,<span class="at">gxyxlVar =</span> <span class="dv">5</span>,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">meanDD =</span> <span class="dv">1</span>,<span class="at">varDD =</span> <span class="dv">5</span>,<span class="at">relAA =</span> <span class="fl">0.1</span>,</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">stageToGenotype =</span> <span class="st">&quot;CET&quot;</span>,</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">nParents =</span> <span class="dv">17</span>, <span class="at">nCrosses =</span> <span class="dv">33</span>, <span class="at">nProgeny =</span> <span class="dv">26</span>,<span class="at">nClonesToNCRP =</span> <span class="dv">3</span>,</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">phenoF1toStage1 =</span> F,<span class="at">errVarPreStage1 =</span> <span class="dv">500</span>,</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">useCurrentPhenoTrain =</span> F, </span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">nCyclesToKeepRecords =</span> <span class="dv">5</span>,</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>                <span class="co"># selCrits are overwritten by runBreedingScheme_wBurnIn</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>                <span class="at">selCritPipeAdv =</span> selCritIID, </span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">selCritPopImprov =</span>  selCritIID) <span class="co"># thus have no actual effect</span></span></code></pre></div>
<ul>
<li><p>Ran the scaled down simulation three times</p>
<ol style="list-style-type: decimal">
<li><p>Set <code>nBLASthreads=60</code> to go as-fast-as-possible… Took 7 mins.</p></li>
<li><p>Set a more reasonable <code>nBLASthreads=5</code> to see how much speed comes from scaled-down sim. size… Took 10 mins.</p></li>
<li><p>Ran <code>nBurnInCycles=15</code> and <code>nPostBurnInCycles=15</code>… took 68 mins.</p></li>
</ol></li>
</ul>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>starttime<span class="ot">&lt;-</span><span class="fu">proc.time</span>()[<span class="dv">3</span>]</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>benchmark_sim2<span class="ot">&lt;-</span><span class="fu">runBreedingScheme_wBurnIn</span>(<span class="at">replication =</span> <span class="dv">1</span>, <span class="at">bsp =</span> bsp,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">nBurnInCycles=</span><span class="dv">2</span>,<span class="at">nPostBurnInCycles=</span><span class="dv">2</span>,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPopPre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPopPost=</span><span class="st">&quot;parentSelCritGEBV&quot;</span>,</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPipePre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPipePost=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">nBLASthreads=</span><span class="dv">60</span>,<span class="at">nThreadsMacs2=</span><span class="dv">60</span>)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>endtime<span class="ot">&lt;-</span><span class="fu">proc.time</span>()[<span class="dv">3</span>]; <span class="fu">print</span>(<span class="fu">paste0</span>((endtime<span class="sc">-</span>starttime)<span class="sc">/</span><span class="dv">60</span>,<span class="st">&quot; mins elapsed.&quot;</span>));</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(benchmark_sim2,<span class="at">file =</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;benchmark_sim2.rds&quot;</span>))</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;6.93941666666667 mins elapsed.&quot;</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>starttime<span class="ot">&lt;-</span><span class="fu">proc.time</span>()[<span class="dv">3</span>]</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>benchmark_sim3<span class="ot">&lt;-</span><span class="fu">runBreedingScheme_wBurnIn</span>(<span class="at">replication =</span> <span class="dv">1</span>, <span class="at">bsp =</span> bsp,</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">nBurnInCycles=</span><span class="dv">2</span>,<span class="at">nPostBurnInCycles=</span><span class="dv">2</span>,</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPopPre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPopPost=</span><span class="st">&quot;parentSelCritGEBV&quot;</span>,</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPipePre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPipePost=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">nBLASthreads=</span><span class="dv">5</span>,<span class="at">nThreadsMacs2=</span><span class="dv">5</span>)</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>endtime<span class="ot">&lt;-</span><span class="fu">proc.time</span>()[<span class="dv">3</span>]; <span class="fu">print</span>(<span class="fu">paste0</span>((endtime<span class="sc">-</span>starttime)<span class="sc">/</span><span class="dv">60</span>,<span class="st">&quot; mins elapsed.&quot;</span>));</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(benchmark_sim3,<span class="at">file =</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;benchmark_sim3.rds&quot;</span>))</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;10.0116333333333 mins elapsed.&quot;</span></span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>starttime<span class="ot">&lt;-</span><span class="fu">proc.time</span>()[<span class="dv">3</span>]</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>benchmark_sim4<span class="ot">&lt;-</span><span class="fu">runBreedingScheme_wBurnIn</span>(<span class="at">replication =</span> <span class="dv">1</span>, <span class="at">bsp =</span> bsp,</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">nBurnInCycles=</span><span class="dv">15</span>,<span class="at">nPostBurnInCycles=</span><span class="dv">15</span>,</span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPopPre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPopPost=</span><span class="st">&quot;parentSelCritGEBV&quot;</span>,</span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPipePre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">selCritPipePost=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">nBLASthreads=</span><span class="dv">5</span>,<span class="at">nThreadsMacs2=</span><span class="dv">5</span>)</span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a>endtime<span class="ot">&lt;-</span><span class="fu">proc.time</span>()[<span class="dv">3</span>]; <span class="fu">print</span>(<span class="fu">paste0</span>((endtime<span class="sc">-</span>starttime)<span class="sc">/</span><span class="dv">60</span>,<span class="st">&quot; mins elapsed.&quot;</span>));</span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(benchmark_sim4,<span class="at">file =</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;benchmark_sim4.rds&quot;</span>))</span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;67.8371833333333 mins elapsed.&quot;</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">readRDS</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;benchmark_sim4.rds&quot;</span>))<span class="sc">$</span>records<span class="sc">$</span>stageOutputs <span class="sc">%&gt;%</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(.,<span class="fu">aes</span>(<span class="at">x=</span>year,<span class="at">y=</span>genValMean,<span class="at">color=</span>stage)) <span class="sc">+</span> </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">15</span>)</span></code></pre></div>
<p><img src="figure/baselineSim.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="run-multiple-scaled-down-sims-in-parallel" class="section level1">
<h1>Run multiple scaled-down sims in parallel</h1>
<p>Set-up 10 iterations of a simulation with 15 pre- and 15-post burn-in cycles.</p>
<p>For a bonus, set-up 10 additional iterations that uses PS (<code>selCritIID</code> the entire 30 cycles).</p>
<p>Made a few changes to the sims so they would exhaust variation (hopefully):</p>
<ul>
<li>Increase to <code>effPopSize=1000</code></li>
<li>Increase stage-specific errors: <code>errVars*3</code></li>
<li>Increase <code>gxyVar</code>, <code>gxlVar</code>, <code>gxyxlVar</code>.</li>
</ul>
<p>Also:</p>
<ul>
<li>Increase to <code>trainingPopCycles=10</code> to use more training data</li>
<li>Set <code>nCyclesToKeepRecords = 30</code> to keep all records</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">screen</span><span class="kw">;</span> </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> shell ~/rocker2.sif<span class="kw">;</span> </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /home/mw489/BreedingSchemeOpt/<span class="kw">;</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(AlphaSimHlpR))</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(tidyverse))</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(genomicMateSelectR))</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>select <span class="ot">&lt;-</span> dplyr<span class="sc">::</span>select</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># This scheme _excludes_ the seedling stage from the simulation.</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>schemeDF<span class="ot">&lt;-</span><span class="fu">read.csv</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;data&quot;</span>,<span class="st">&quot;baselineScheme - IITA.csv&quot;</span>), </span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">header =</span> T, <span class="at">stringsAsFactors =</span> F) <span class="sc">%&gt;%</span> </span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>PlantsPerPlot) <span class="sc">%&gt;%</span> </span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">trainingPopCycles=</span><span class="dv">10</span>,</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">nEntries=</span><span class="fu">ceiling</span>(nEntries<span class="sc">/</span><span class="dv">3</span>),</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">nChks=</span><span class="fu">ceiling</span>(nChks<span class="sc">/</span><span class="dv">3</span>),</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">errVars=</span>errVars<span class="sc">*</span><span class="dv">3</span>)</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>bsp<span class="ot">&lt;-</span><span class="fu">specifyBSP</span>(<span class="at">schemeDF =</span> schemeDF,</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>                <span class="at">nChr =</span> <span class="dv">3</span>,<span class="at">effPopSize =</span> <span class="dv">1000</span>,<span class="at">quickHaplo =</span> F,</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>                <span class="at">segSites =</span> <span class="dv">200</span>, <span class="at">nQTL =</span> <span class="dv">30</span>, <span class="at">nSNP =</span> <span class="dv">100</span>, <span class="at">genVar =</span> <span class="dv">200</span>,</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>                <span class="at">gxeVar =</span> <span class="cn">NULL</span>, <span class="at">gxyVar =</span> <span class="dv">20</span>, <span class="at">gxlVar =</span> <span class="dv">15</span>,<span class="at">gxyxlVar =</span> <span class="dv">10</span>,</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>                <span class="at">meanDD =</span> <span class="dv">1</span>,<span class="at">varDD =</span> <span class="dv">5</span>,<span class="at">relAA =</span> <span class="fl">0.1</span>,</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>                <span class="at">stageToGenotype =</span> <span class="st">&quot;CET&quot;</span>,</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>                <span class="at">nParents =</span> <span class="dv">17</span>, <span class="at">nCrosses =</span> <span class="dv">33</span>, <span class="at">nProgeny =</span> <span class="dv">26</span>,<span class="at">nClonesToNCRP =</span> <span class="dv">3</span>,</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>                <span class="at">phenoF1toStage1 =</span> F,<span class="at">errVarPreStage1 =</span> <span class="dv">500</span>,</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>                <span class="at">useCurrentPhenoTrain =</span> F, </span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>                <span class="at">nCyclesToKeepRecords =</span> <span class="dv">30</span>,</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>                <span class="co"># selCrits are overwritten by runBreedingScheme_wBurnIn</span></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>                <span class="at">selCritPipeAdv =</span> selCritIID, </span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>                <span class="at">selCritPopImprov =</span>  selCritIID) <span class="co"># thus have no actual effect</span></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>benchmark_sims<span class="ot">&lt;-</span><span class="fu">crossing</span>(<span class="at">replication=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,</span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>                         <span class="at">postBurnIn=</span><span class="fu">c</span>(<span class="st">&quot;parentSelCritGEBV&quot;</span>,<span class="st">&quot;selCritIID&quot;</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(postBurnIn)</span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>benchmark_sims <span class="sc">%&gt;%</span> rmarkdown<span class="sc">::</span><span class="fu">paged_table</span>()</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["replication"],"name":[1],"type":["int"],"align":["right"]},{"label":["postBurnIn"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"1","2":"parentSelCritGEBV"},{"1":"2","2":"parentSelCritGEBV"},{"1":"3","2":"parentSelCritGEBV"},{"1":"4","2":"parentSelCritGEBV"},{"1":"5","2":"parentSelCritGEBV"},{"1":"6","2":"parentSelCritGEBV"},{"1":"7","2":"parentSelCritGEBV"},{"1":"8","2":"parentSelCritGEBV"},{"1":"9","2":"parentSelCritGEBV"},{"1":"10","2":"parentSelCritGEBV"},{"1":"1","2":"selCritIID"},{"1":"2","2":"selCritIID"},{"1":"3","2":"selCritIID"},{"1":"4","2":"selCritIID"},{"1":"5","2":"selCritIID"},{"1":"6","2":"selCritIID"},{"1":"7","2":"selCritIID"},{"1":"8","2":"selCritIID"},{"1":"9","2":"selCritIID"},{"1":"10","2":"selCritIID"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Run 10 sims in parallel, each sim gets 10 additional threads for BLAS to speed GP.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>starttime<span class="ot">&lt;-</span><span class="fu">proc.time</span>()[<span class="dv">3</span>]</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(furrr); <span class="fu">plan</span>(multisession, <span class="at">workers =</span> <span class="dv">10</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize=</span><span class="sc">+</span><span class="cn">Inf</span>); <span class="fu">options</span>(<span class="at">future.rng.onMisuse=</span><span class="st">&quot;ignore&quot;</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>benchmark_sims <span class="sc">%&lt;&gt;%</span> </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sim=</span><span class="fu">future_map2</span>(replication,postBurnIn,</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>                         <span class="sc">~</span><span class="fu">runBreedingScheme_wBurnIn</span>(<span class="at">replication =</span> .x, </span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">bsp =</span> bsp,</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">nBurnInCycles=</span><span class="dv">15</span>,<span class="at">nPostBurnInCycles=</span><span class="dv">15</span>,</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">selCritPopPre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">selCritPopPost=</span>.y,</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">selCritPipePre=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">selCritPipePost=</span><span class="st">&quot;selCritIID&quot;</span>,</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">nBLASthreads=</span><span class="dv">10</span>,<span class="at">nThreadsMacs2=</span><span class="dv">10</span>)))</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>endtime<span class="ot">&lt;-</span><span class="fu">proc.time</span>()[<span class="dv">3</span>]; <span class="fu">print</span>(<span class="fu">paste0</span>((endtime<span class="sc">-</span>starttime)<span class="sc">/</span><span class="dv">60</span>,<span class="st">&quot; mins elapsed.&quot;</span>));</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(benchmark_sims,<span class="at">file =</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;benchmark_sims5.rds&quot;</span>))</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>(sequential)</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;499.316183333333 mins elapsed.&quot;</span></span></code></pre></div>
<p>Took ~8.3 hrs to run.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>sims<span class="ot">&lt;-</span><span class="fu">readRDS</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;output&quot;</span>,<span class="st">&quot;benchmark_sims5.rds&quot;</span>))</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>sims <span class="sc">%&gt;%</span> </span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sim=</span><span class="fu">map</span>(sim,<span class="sc">~</span>.<span class="sc">$</span>records<span class="sc">$</span>stageOutputs)) <span class="sc">%&gt;%</span> </span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(sim) <span class="sc">%&gt;%</span> </span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(stage<span class="sc">==</span><span class="st">&quot;CET&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">rep=</span><span class="fu">paste0</span>(postBurnIn,replication)) <span class="sc">%&gt;%</span> </span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  <span class="co">#as.character(replication)) %&gt;% </span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  <span class="co">#stage=factor(stage,levels = c(&quot;F1&quot;,&quot;CET&quot;,&quot;PYT&quot;,&quot;AYT&quot;,&quot;UYT&quot;)))</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(.,<span class="fu">aes</span>(<span class="at">x=</span>year,<span class="at">y=</span>genValMean,<span class="at">color=</span>postBurnIn, <span class="at">group=</span>rep)) <span class="sc">+</span> </span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">15</span>)</span></code></pre></div>
<p><img src="figure/baselineSim.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Plot the mean genetic value (y-axis) of the “CET” stage for all 20 simulations (10 with <code>selCritIID</code> PS and 10 with <code>parentSelCritGEBV</code> GS used after the first 15 cycles of PS). <strong>NOTE:</strong> In retrospect, it is not entirely fair to compare the post-burn-in PS (blue) and GS (red). Next time, I will set-up a <code>runBreedingScheme</code> function that simulations a single population pre-burn-in and then diverges that population post-burn-in. A proper simulation would diver</p>
</div>
<div id="questions-and-next-steps" class="section level1">
<h1>Questions and next steps</h1>
<ol style="list-style-type: decimal">
<li><p>Correct <code>runBreedingScheme_wBurnIn</code> to diverge single pop after burn-in.</p></li>
<li><p>I am not yet seeing a notable effect of switching from PS to GS after 15 cycles. What parameters need to be changed?</p></li>
<li><p>How to handle all the breeding simulation options, e.g.:</p>
<ul>
<li><p>Ne, <code>entryToChkRatio</code>, etc.</p></li>
<li><p>Similar approach to with plot size…. simulate a spectrum and check for an effect?</p></li>
</ul></li>
</ol>
<!-- -->
<ol start="4" style="list-style-type: decimal">
<li>Develop and test a errVar-plotSize scaling function</li>
<li>Complete a baseline simulation</li>
<li>Test alternative <em>mate selection</em> scenarios, etc.</li>
</ol>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.1.0 (2021-05-18)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] genomicMateSelectR_0.2.0 forcats_0.5.1            stringr_1.4.0           
 [4] purrr_0.3.4              readr_2.0.1              tidyr_1.1.3             
 [7] tibble_3.1.3             ggplot2_3.3.5            tidyverse_1.3.1         
[10] AlphaSimHlpR_0.2.1       dplyr_1.0.7              AlphaSimR_1.0.3         
[13] R6_2.5.0                 workflowr_1.6.2         

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7        here_1.0.1        lubridate_1.7.10  assertthat_0.2.1 
 [5] rprojroot_2.0.2   digest_0.6.27     utf8_1.2.2        cellranger_1.1.0 
 [9] backports_1.2.1   reprex_2.0.1      evaluate_0.14     highr_0.9        
[13] httr_1.4.2        pillar_1.6.2      rlang_0.4.11      readxl_1.3.1     
[17] rstudioapi_0.13   whisker_0.4       jquerylib_0.1.4   rmarkdown_2.10   
[21] labeling_0.4.2    munsell_0.5.0     broom_0.7.9       compiler_4.1.0   
[25] httpuv_1.6.1      modelr_0.1.8      xfun_0.25         pkgconfig_2.0.3  
[29] htmltools_0.5.1.1 tidyselect_1.1.1  fansi_0.5.0       crayon_1.4.1     
[33] tzdb_0.1.2        dbplyr_2.1.1      withr_2.4.2       later_1.2.0      
[37] grid_4.1.0        jsonlite_1.7.2    gtable_0.3.0      lifecycle_1.0.0  
[41] DBI_1.1.1         git2r_0.28.0      magrittr_2.0.1    scales_1.1.1     
[45] cli_3.0.1         stringi_1.7.3     farver_2.1.0      fs_1.5.0         
[49] promises_1.2.0.1  xml2_1.3.2        bslib_0.2.5.1     ellipsis_0.3.2   
[53] generics_0.1.0    vctrs_0.3.8       tools_4.1.0       glue_1.4.2       
[57] hms_1.1.0         yaml_2.2.1        colorspace_2.0-2  rvest_1.0.1      
[61] knitr_1.33        haven_2.4.3       sass_0.4.0       </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
