---
title: "Example of AlphaSimHlpR simulation with a burn-in period"
author: "Marnin Wolfe"
date: "2021-Sep-17"
output: 
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, 
                      tidy='styler', 
                      tidy.opts=list(strict=FALSE,width.cutoff=100), 
                      highlight=TRUE)
```

Below, I show an example of how to run a simulation that includes a burn-in and post burn-in phase.

1. Use `runBurnInScheme()` function (not yet in `AlphaSimHlpR`; sourced from `code/runBurnInScheme.R`) to initiate a simulation with phenotypic selection.
2. Use `runSchemesPostBurnIn()` with a new `bsp` (sourced from `code/runSchemesPostBurnIn.R`) to continue simulating restarting the previously started (burnt-in) simulations created in Step 1 but potentially with new `bsp` settings.

**Install the version of `AlphaSimHlpR` that I have been working on.**
```{r, eval=F}
# To install the latest version
devtools::install_github("wolfemd/AlphaSimHlpR", ref = 'master')
```

See full [function reference manual here](https://wolfemd.github.io/AlphaSimHlpR/index.html) in my forked-repo of `AlphaSimHlpR`

# A small example

Test the code with a small example. Source functions *not yet* included in `AlphaSimHlpR` from `code/` directory.

```{r,eval=T}
suppressMessages(library(AlphaSimHlpR))
suppressMessages(library(tidyverse))
suppressMessages(library(genomicMateSelectR))
select <- dplyr::select

schemeDF<-read.csv(here::here("data","baselineScheme - Test.csv"), 
                   header = T, stringsAsFactors = F)

bsp<-specifyBSP(schemeDF = schemeDF,
                nChr = 3,effPopSize = 100,quickHaplo = F,
                segSites = 400, nQTL = 40, nSNP = 100, genVar = 40,
                gxeVar = NULL, gxyVar = 15, gxlVar = 10,gxyxlVar = 5,
                meanDD = 0.5,varDD = 0.01,relAA = 0.5,
                stageToGenotype = "PYT",
                nParents = 10, nCrosses = 4, nProgeny = 50,nClonesToNCRP = 3,
                phenoF1toStage1 = T,errVarPreStage1 = 500,
                useCurrentPhenoTrain = F, 
                nCyclesToKeepRecords = 30,
                nTrainPopCycles=6,
                nYrsAsCandidates=1,
                maxTrainingPopSize=500)

source(here::here("code","runBurnInSchemes.R"))
source(here::here("code","runSchemesPostBurnIn.R"))
```

I created a CSV to specify a data.frame `schemeDF` defining stage-specific breeding scheme inputs.

```{r}
schemeDF %>% rmarkdown::paged_table()
```

Run 10 cycles of burn-in simulation. 

The default is to set-up phenotypic selection only during burn-in.
```{r run test burn-in sims, eval=F}
burnInSims<-runBurnInSchemes(bsp = bsp,
                             nBurnInCycles=10,
                             selCritPop="parentSelCritBLUP",
                             selCritPipe="selCritIID",
                             iniFunc="initializeScheme",
                             productFunc="productPipeline",
                             popImprovFunc="popImprovByParentSel",
                             nReplications=4,nSimCores=4,
                             nBLASthreads=1,nThreadsMacs2=1)
saveRDS(burnInSims,file = here::here("output","test_burnInSims_2021Sep17.rds"))
```

Two sets of post burn-in simulations, both with same `bsp` overall. 

(1) continue with phenotypic selection, _no change_. 

(2) Switch to `parentSelCritGEBV`.

**NOTE:** Below, switch (by default) to `productPipelinePostBurnIn` for the product advancement pipeline _and_ it's corresponding selection criteria `productSelCritBLUP`

```{r post burn-in - same bsp, eval=F}
burnInSims<-readRDS(file = here::here("output","test_burnInSims_2021Sep17.rds"))
postBurnIn_PS<-runSchemesPostBurnIn(simulations = burnInSims,
                                    nPostBurnInCycles=10,
                                    selCritPop="parentSelCritBLUP",
                                    selCritPipe="productSelCritBLUP",
                                    productFunc="productPipelinePostBurnIn",
                                    popImprovFunc="popImprovByParentSel",
                                    nSimCores=4,
                                    nBLASthreads=1)
saveRDS(postBurnIn_PS,file = here::here("output","test_burnInSims_PS_2021Sep17.rds"))
#postBurnIn_PS$SimOutput[[1]]$records$stageOutputs
postBurnIn_GS<-runSchemesPostBurnIn(simulations = burnInSims,
                                    nPostBurnInCycles=10,
                                    selCritPop="parentSelCritGEBV",
                                    selCritPipe="productSelCritBLUP",
                                    productFunc="productPipelinePostBurnIn",
                                    popImprovFunc="popImprovByParentSel",
                                    nSimCores=4,
                                    nBLASthreads=1)
saveRDS(postBurnIn_GS,file = here::here("output","test_burnInSims_GS_2021Sep17.rds"))
#postBurnIn_GS$SimOutput[[1]]$records$stageOutputs
```

```{r, eval=T}
forSimPlot<-readRDS(file = here::here("output","test_burnInSims_GS_2021Sep17.rds")) %>% 
  mutate(PostBurnIn="GS") %>% 
  bind_rows(readRDS(file = here::here("output","test_burnInSims_PS_2021Sep17.rds")) %>% 
              mutate(PostBurnIn="PS")) %>% 
  unnest_wider(SimOutput) %>% 
  select(SimRep,PostBurnIn,records) %>% 
  unnest_wider(records) %>% 
  select(SimRep,PostBurnIn,stageOutputs) %>% 
  unnest() %>% 
  filter(stage=="F1") %>% 
  mutate(YearPostBurnIn=year-10)
 
library(patchwork)
meanGplot<-forSimPlot %>% 
  mutate(SimRep=paste0(PostBurnIn,SimRep)) %>% 
  group_by(PostBurnIn,YearPostBurnIn,year,stage) %>% 
  summarize(meanGenMean=mean(genValMean),
            seGenMean=sd(genValMean)/sqrt(n())) %>% 
  ggplot(.,aes(x=YearPostBurnIn)) +
  geom_ribbon(aes(ymin = meanGenMean - seGenMean, 
                  ymax = meanGenMean + seGenMean,
                  fill = PostBurnIn), 
              alpha=0.75) + 
  geom_line(aes(y = meanGenMean, color=PostBurnIn))
sdGplot<-forSimPlot %>% 
  mutate(SimRep=paste0(PostBurnIn,SimRep)) %>% 
  group_by(PostBurnIn,YearPostBurnIn,year,stage) %>% 
  summarize(meanGenSD=mean(genValSD),
            seGenSD=sd(genValSD)/sqrt(n())) %>% 
  ggplot(.,aes(x=YearPostBurnIn)) +
  geom_ribbon(aes(ymin = meanGenSD - seGenSD, 
                  ymax = meanGenSD + seGenSD,
                  fill = PostBurnIn), 
              alpha=0.75) + 
  geom_line(aes(y = meanGenSD, color=PostBurnIn))
(meanGplot | sdGplot) + patchwork::plot_layout(guides = 'collect') & 
  theme_bw() & geom_vline(xintercept = 0, color='darkred')
```

# Change the VDP

Try post burn-in sims with an altered VDP (specified in the `bsp`). 

_Do not_ change pop. genetic / genomic parameters. 

Try removing middle "AYT" stage of the example `schemeDF`.
```{r post burn-in - NEW bsp, eval=F}
suppressMessages(library(AlphaSimHlpR))
suppressMessages(library(tidyverse))
suppressMessages(library(genomicMateSelectR))
select <- dplyr::select

schemeDF<-read.csv(here::here("data","baselineScheme - Test.csv"), 
                   header = T, stringsAsFactors = F)

source(here::here("code","runSchemesPostBurnIn.R"))

burnInSims<-readRDS(file = here::here("output","test_burnInSims_2021Sep17.rds"))
newBSP<-specifyBSP(schemeDF = schemeDF %>% filter(stageNames!="AYT"),
                   nChr = 3,effPopSize = 100,quickHaplo = F,
                   segSites = 400, nQTL = 40, nSNP = 100, genVar = 40,
                   gxeVar = NULL, gxyVar = 15, gxlVar = 10,gxyxlVar = 5,
                   meanDD = 0.5,varDD = 0.01,relAA = 0.5,
                   stageToGenotype = "PYT",
                   nParents = 10, nCrosses = 4, nProgeny = 50,nClonesToNCRP = 3,
                   phenoF1toStage1 = T,errVarPreStage1 = 500,
                   useCurrentPhenoTrain = F,
                   nCyclesToKeepRecords = 30,
                   nTrainPopCycles=6,
                   nYrsAsCandidates=1,
                   maxTrainingPopSize=500)

postBurnIn_PS_newBSP<-runSchemesPostBurnIn(simulations = burnInSims,
                                           newBSP = newBSP,
                                           nPostBurnInCycles=10,
                                           selCritPop="parentSelCritBLUP",
                                           selCritPipe="productSelCritBLUP",
                                           productFunc="productPipelinePostBurnIn",
                                           popImprovFunc="popImprovByParentSel",
                                           nSimCores=4,
                                           nBLASthreads=1)
saveRDS(postBurnIn_PS_newBSP,file = here::here("output","test_burnInSims_PS_noAYT_2021Sep17.rds"))

postBurnIn_GS_newBSP<-runSchemesPostBurnIn(simulations = burnInSims,
                                           newBSP = newBSP,
                                           nPostBurnInCycles=10,
                                           selCritPop="parentSelCritGEBV",
                                           selCritPipe="productSelCritBLUP",
                                           productFunc="productPipelinePostBurnIn",
                                           popImprovFunc="popImprovByParentSel",
                                           nSimCores=4,
                                           nBLASthreads=1)
saveRDS(postBurnIn_GS_newBSP,file = here::here("output","test_burnInSims_GS_noAYT_2021Sep17.rds"))

```
```{r, eval=T}
forSimPlot<-readRDS(file = here::here("output","test_burnInSims_GS_2021Sep17.rds")) %>% 
  mutate(PostBurnIn="GS") %>% 
  bind_rows(readRDS(file = here::here("output","test_burnInSims_PS_2021Sep17.rds")) %>% 
              mutate(PostBurnIn="PS")) %>% 
  bind_rows(readRDS(file = here::here("output","test_burnInSims_PS_noAYT_2021Sep17.rds")) %>% 
              mutate(PostBurnIn="PS_noAYT")) %>% 
  bind_rows(readRDS(file = here::here("output","test_burnInSims_GS_noAYT_2021Sep17.rds")) %>% 
              mutate(PostBurnIn="GS_noAYT")) %>% 
  unnest_wider(SimOutput) %>% 
  select(SimRep,PostBurnIn,records) %>% 
  unnest_wider(records) %>% 
  select(SimRep,PostBurnIn,stageOutputs) %>% 
  unnest() %>% 
  filter(stage=="F1") %>% 
  mutate(YearPostBurnIn=year-10)
 
library(patchwork)
meanGplot<-forSimPlot %>% 
  mutate(SimRep=paste0(PostBurnIn,SimRep)) %>% 
  group_by(PostBurnIn,YearPostBurnIn,year,stage) %>% 
  summarize(meanGenMean=mean(genValMean),
            seGenMean=sd(genValMean)/sqrt(n())) %>% 
  ggplot(.,aes(x=YearPostBurnIn)) +
  geom_ribbon(aes(ymin = meanGenMean - seGenMean, 
                  ymax = meanGenMean + seGenMean,
                  fill = PostBurnIn), 
              alpha=0.75) + 
  geom_line(aes(y = meanGenMean, color=PostBurnIn))
sdGplot<-forSimPlot %>% 
  mutate(SimRep=paste0(PostBurnIn,SimRep)) %>% 
  group_by(PostBurnIn,YearPostBurnIn,year,stage) %>% 
  summarize(meanGenSD=mean(genValSD),
            seGenSD=sd(genValSD)/sqrt(n())) %>% 
  ggplot(.,aes(x=YearPostBurnIn)) +
  geom_ribbon(aes(ymin = meanGenSD - seGenSD, 
                  ymax = meanGenSD + seGenSD,
                  fill = PostBurnIn), 
              alpha=0.75) + 
  geom_line(aes(y = meanGenSD, color=PostBurnIn))
(meanGplot | sdGplot) + patchwork::plot_layout(guides = 'collect') & 
  theme_bw() & geom_vline(xintercept = 0, color='darkred')
```

# Run full-scale burn-in sims

Previously, I used an empirical approach to estimate TrialType-specific error variances in terms of the IITA selection index (SELIND). See that analysis [here](https://wolfemd.github.io/IITA_2021GS/inputsForSimulationV2.html). 

**TO DO:** Need to run full-scale burn-in simulations for each breeding program. 

20 burn-in cycles to match examples by EiB.

-   Genome / Pop specs

    -   18 chrom,

    -   Ne = 1000,

    -   nSNP = 300 SNP/chrom (matches EiB examples)

    -   nQTLperChr = 1000

    -   nSegSites = 2000

-   Genetic architecture and Error variance

    -   `genVar = 750` and stage-specific `errVar` input from [here](https://wolfemd.github.io/IITA_2021GS/inputsForSimulationV2.html)

        -   The max estimated `errVar` was for CET at \~3500,

        -   so a `genVar` of 750 is to set up a entry level h2 around 0.2

    -   `meanDD = 0.3` and `varDD = 0.05` 
    
        -   or `MeanDD=0.23` and `VarDD=0.06`, based loosely on [this estimate and note](https://wolfemd.github.io/IITA_2021GS/inputsForSimulationV2.html#A_note_on_inputs_for_dominance).

    -   Var(GxYr) == Var(G), again matching EiB example

        -   What about GxL and GxLxYr?

```{r}
read.csv(here::here("data","baselineScheme - IITA.csv"), 
                   header = T, stringsAsFactors = F) %>% 
  select(-errVars,-PlantsPerPlot) %>% 
  left_join(readRDS(here::here("data","siErrorVarEst_byTrialType_directApproach_2021Aug25.rds")) %>% 
              select(-VarEsts) %>% 
              rename(errVars=siErrorVarEst)) %>% 
  select(-TrialType) %>% 
  mutate(trainingPopCycles=20) %>% 
  rmarkdown::paged_table()
```

-   Breeding Scheme (`schemeDF` printed above)

    -   Skips SDN stage. Is there an UYT2 (second year of UYT) to sim?

    -   `phenoF1toStage1 = FALSE`

    -   Population Improvement

        -   `nParents = 50, nCrosses = 100, nProgeny = 25,nClonesToNCRP = 3`
        -   Or `nParents = 100, nCrosses = 250, nProgeny = 10,nClonesToNCRP = 3` (EiB example)

-   Additional Settings

    -   `nCyclesToKeepRecords = 30` (all)... what effect does this actually have? Just on storage of output?

    -   `trainingPopCycles = 15`...

        -   means 15 years of each stage used in each prediction...

        -   What about an alternative: set a fixed TP size e.g. 5000 clones.

            -   This might be faster since the number of clones / dimension of kinship matrix is primary slow point.


**Run multiple versions of an burn-in simulation for 20 cycles.** 



# Next steps

1.  Complete full-scale burn-in simulations 

2.  [Conduct a baseline post burn-in GS vs. Conv. simulation](baselineSim.html)

3.  Burn-in and baseline simulations for National programs (NaCRRI, TARI, NRCRI, EMBRAPA). 
  _Still need input re: selection index weights and current program structure._
  
4.  Begin the actually interesting simulations

    -   Optimize budgets

    -   Compare alternative VDPs

    -   Test *mate selection*, *optimal contributions* and ultimately optimizing *mating plans*.
