---
title: "Simulate genomic mate selection"
author: "Marnin Wolfe"
date: "2021-09-15"
output: 
  workflowr::wflow_html:
    code_folding: hide
    toc: true
editor_options:
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, 
                      tidy='styler', 
                      tidy.opts=list(strict=FALSE,width.cutoff=100), 
                      highlight=TRUE)
```

```{bash environ, eval=F}
# 0) Pull a singularity image with OpenBLAS enabled R + tidyverse from rocker/
# singularity pull ~/rocker2.sif docker://rocker/tidyverse:latest;

# 1) start a screen shell 
screen; 
# 2) reserve interactive slurm
salloc -n 20 --mem 60G;
# 3) start the singularity Linux shell inside that
singularity shell ~/rocker2.sif; 
# Project directory, so R will use as working dir.
cd /home/mw489/BreedingSchemeOpt/;
# 3) Start R
export OMP_NUM_THREADS=1;
R
```

# Small example

```{r}
suppressMessages(library(AlphaSimHlpR))
suppressMessages(library(tidyverse))
suppressMessages(library(genomicMateSelectR))
select <- dplyr::select

schemeDF<-read.csv(here::here("data","baselineScheme - Test.csv"), 
                   header = T, stringsAsFactors = F)

source(here::here("code","runSchemesPostBurnIn.R"))

simulations<-readRDS(here::here("output","burnIn_test.rds"))

newBSP<-specifyBSP(schemeDF = schemeDF,
                   nChr = 3,effPopSize = 100,quickHaplo = F,
                   segSites = 400, nQTL = 40, nSNP = 100, genVar = 40,
                   gxeVar = NULL, gxyVar = 15, gxlVar = 10,gxyxlVar = 5,
                   meanDD = 0.5,varDD = 0.01,relAA = 0.5,
                   stageToGenotype = "PYT",
                   nParents = 10, nCrosses = 4, nProgeny = 50,nClonesToNCRP = 3,
                   phenoF1toStage1 = T,errVarPreStage1 = 500,
                   useCurrentPhenoTrain = F, 
                   nCyclesToKeepRecords = 30,
                   nTrainPopCycles=6,
                   nYrsAsCandidates=2,
                   maxTrainingPopSize=500,
                   crossSelCrit="MeanBV",
                   modelType="A")
```

```{r gms - model A - MeanBV - 10 parents}

start<-proc.time()[3]
testSims<-runSchemesPostBurnIn(simulations = simulations %>% slice(1),
                               newBSP=newBSP,
                               nPostBurnInCycles=5,
                               selCritPop="genomicMateSelCrit",
                               selCritPipe="selCritIID",
                               productFunc="productPipeline",
                               popImprovFunc="popImprovByMateSel",
                               nSimCores=4,
                               nBLASthreads=1)
end<-proc.time()[3]; timebsp<-end-start;
print(paste0((timebsp)/60," mins elapsed"))
# Error: Problem with `mutate()` column `SimOutput`. 
# ℹ `SimOutput = future_map2(...)`. 
# x invalid class “Pop” object: nInd!=length(fixEff) 
```
```{r}
#saveRDS(testSims,file = here::here("output","testSims_GMS_ModelA_MeanBV_2021Sep15.rds"))
```

# debug nInd!=length(fixEff) problem

```{r}
simulations %<>% slice(1)
selCritPop="genomicMateSelCrit";
selCritPipe="selCritIID";
productFunc="productPipeline";
popImprovFunc="popImprovByMateSel";
nSimCores=4;
nBLASthreads=1
nPostBurnInCycles=5
# require(furrr); plan(multisession, workers = nSimCores)
# options(future.globals.maxSize=+Inf); options(future.rng.onMisuse="ignore")

  # simulations<-simulations %>%
  #   mutate(SimOutput=future_map2(SimRep,burnInSim,function(SimRep,burnInSim,...){
      # debug
      burnInSim<-simulations$burnInSim[[1]]; SimRep<-simulations$SimRep[[1]]
      if(!is.null(nBLASthreads)) { RhpcBLASctl::blas_set_num_threads(nBLASthreads) }
      cat("******", SimRep, "\n")

      # This CONTINUES where previous sims left off
      ## no initialize step
      ## Keep burn-in stage sim params "SP"
      SP<-burnInSim$SP
      ## specify a potentially new bsp object
      ## (keep checks stored in burn-in stage's bsp)
      if(!is.null(newBSP)){
        bsp<-newBSP; bsp$checks<-burnInSim$bsp$checks
      } else { bsp<-burnInSim$bsp }
      ## 'historical' records from burn-in
      records<-burnInSim$records
      ## override burn-in specified product and population improvement funcs
      bsp[["productPipeline"]] <- get(productFunc)
      bsp[["populationImprovement"]] <- get(popImprovFunc)
      bsp[["selCritPipeAdv"]] <- get(selCritPipe)
      bsp[["selCritPopImprov"]] <- get(selCritPop)

      # Post burn-in cycles
      cat("\n"); cat("Post burn-in cycles"); cat("\n")
      for (cycle in 1:nPostBurnInCycles){
        cat(cycle, " ")
        records <- bsp$productPipeline(records, bsp, SP)
        records <- bsp$populationImprovement(records, bsp, SP)
      }
```
```{r}
records$F1
```
```{r}
records$F1@fixEff %>% table
```
```{r}
records$stageOutputs %>% count(year)
```
```{r}
validObject(records$F1)
```
```{r}
trainRec <- records
  if (!bsp$useCurrentPhenoTrain){
    for (stage in 1+1:bsp$nStages){
      trainRec[[stage]] <- trainRec[[stage]][-length(trainRec[[stage]])]
    }
  }

  if(bsp$stageToGenotype=="F1"){
    NrecentProgenySelCands<-(bsp$nProgeny*bsp$nCrosses)*bsp$nYrsAsCandidates
    candidates<-records$F1@id %>% tail(.,n = NrecentProgenySelCands)
  } else {
    candidates<-records[[bsp$stageToGenotype]] %>%
      tail(.,n=bsp$nYrsAsCandidates) %>%
      map_df(.,bind_rows) %$%
      unique(id) %>%
      # exclude checks
      setdiff(.,bsp$checks@id)
  }

  phenotypedLines<-trainRec[bsp$stageNames] %>%
    map(.,~tail(.,n = bsp$nTrainPopCycles)) %>%
    map_df(.,bind_rows) %$%
    unique(id)

  phenotypedLines_notSelCands<-setdiff(phenotypedLines,candidates)
  ## maxTPsize is lesser of specified 'maxTrainingPopSize' and actual number of phenotyped lines not considered selection candidates
  maxTPsize<-min(bsp$maxTrainingPopSize,length(phenotypedLines_notSelCands))
  ## Make sure checks ARE included
  if(!is.null(bsp$checks)){
    # sample from the list of non-selection candidates that also are NOT checks
    trainingpop<-sample(setdiff(phenotypedLines_notSelCands,bsp$checks@id),
                        size = maxTPsize, replace = F) %>%
      # include the checks
      c(.,bsp$checks@id) %>%
      # vanity: order the ids
      .[order(as.integer(.))]
  } else {
    trainingpop<-sample(phenotypedLines_notSelCands,
                        size = maxTPsize, replace = F) %>%
      .[order(as.integer(.))]
  }

  crit <- bsp$selCritPopImprov(trainRec, candidates, trainingpop, bsp, SP)
  crit
```
```{r}
crit %>% 
  slice_max(order_by = crossSelCrit,
            n = bsp$nCrosses,with_ties = F)
```


```{r}
# select the top nCrosses
  crossingPlan<-crit %>%
    slice_max(order_by = crossSelCrit,
              n = bsp$nCrosses) %>%
    select(sireID,damID) %>%
    as.matrix
crossingPlan
```

```{r}
# extract a pop-object of those parents
  parents <- records$F1[crossingPlan %>% as.vector %>% unique]
parents
```

```{r}
# make crosses
  progeny <- makeCross(pop = parents,
                       crossPlan = crossingPlan,
                       nProgeny = bsp$nProgeny, simParam=SP)
progeny
```
Now the `progeny` pop object has 250 indivs instead of 200 expected, because of the extra cross. 

```{r}
length(progeny@fixEff)
```
And the next line induces the problem. `bsp$nSeeds` says there should be 200 not 250 elements to `progeny@fixEff`...

```{r}
progeny@fixEff <- rep(as.integer(max(records$stageOutputs$year) + 1), bsp$nSeeds)
progeny
```
```{r}
validObject(progeny)
```
Bam!

```{r}
parentsUsed <- unique(c(progeny@mother, progeny@father))
  stgCyc <- sapply(parentsUsed, AlphaSimHlpR:::whereIsID, records=records)
  stgCyc <- table(stgCyc[1,], stgCyc[2,])
  strtStgOut <- nrow(records$stageOutputs) - bsp$nStages - 1
  for (i in 1:nrow(stgCyc)){
    stage <- as.integer(rownames(stgCyc)[i])
    records$stageOutputs$nContribToPar[[strtStgOut + stage]] <- tibble(cycle=as.integer(colnames(stgCyc)), nContribToPar=stgCyc[i,])
  }
  records$F1 <- c(records$F1, progeny)
```


```{r}
# # Finalize the stageOutputs
#       records <- AlphaSimHlpR:::lastCycStgOut(records, bsp, SP)
# 
#       return(list(records=records,
#                   bsp=bsp,
#                   SP=SP))
#     },
#     nPostBurnInCycles=nPostBurnInCycles,
#     selCritPop=selCritPop,
#     selCritPipe=selCritPipe,
#     productFunc=productFunc,
#     popImprovFunc=popImprovFunc,
#     nBLASthreads=nBLASthreads,
#     newBSP=newBSP))
#   plan(sequential)
```


## Plot results
```{r plot test sims}
postBurnInGMS<-readRDS(here::here("output","postBurnInGMS_test_DirDom_UCparent.rds"))
postBurnInGS<-readRDS(here::here("output","postBurnInGS_test.rds"))
postBurnInPS<-readRDS(here::here("output","postBurnIn_test.rds"))

forSimPlot<-postBurnInGMS %>% 
  mutate(PostBurnIn="GMS") %>% 
  bind_rows(postBurnInGS %>% 
              mutate(PostBurnIn="GS")) %>% 
  bind_rows(postBurnInPS %>% 
              mutate(PostBurnIn="PS")) %>% 
  unnest_wider(SimOutput) %>% 
  select(SimRep,PostBurnIn,records) %>% 
  unnest_wider(records) %>% 
  select(SimRep,PostBurnIn,stageOutputs) %>% 
  unnest() %>% 
  filter(stage=="F1") %>% 
  mutate(YearPostBurnIn=year-10)
 
library(patchwork)
meanGplot<-forSimPlot %>% 
  group_by(PostBurnIn,YearPostBurnIn,year,stage) %>% 
  summarize(meanGenMean=mean(genValMean),
            seGenMean=sd(genValMean)/sqrt(n())) %>% 
  ggplot(.,aes(x=YearPostBurnIn)) +
  geom_ribbon(aes(ymin = meanGenMean - seGenMean, 
                  ymax = meanGenMean + seGenMean,
                  fill = PostBurnIn), 
              alpha=0.75) + 
  geom_line(aes(y = meanGenMean, color=PostBurnIn))
sdGplot<-forSimPlot %>% 
  group_by(PostBurnIn,YearPostBurnIn,year,stage) %>% 
  summarize(meanGenSD=mean(genValSD),
            seGenSD=sd(genValSD)/sqrt(n())) %>% 
  ggplot(.,aes(x=YearPostBurnIn)) +
  geom_ribbon(aes(ymin = meanGenSD - seGenSD, 
                  ymax = meanGenSD + seGenSD,
                  fill = PostBurnIn), 
              alpha=0.75) + 
  geom_line(aes(y = meanGenSD, group=PostBurnIn))
(meanGplot | sdGplot) + patchwork::plot_layout(guides = 'collect') & 
  theme_bw() & geom_vline(xintercept = 0, color='darkred')
```

# Another set of (small) GMS sims

Obviously, my first attempt used what I'm guessing is far too much weight on the predicted cross-variance. 

Whatever happened, I need to fiddle with the parameters to get a sense for what is happening.
```{r}
# nParents --> n CrossesToPredict --> compute time estimate
# 100=5050 (90 mins estimated)
# 50=1275 (23 mins estimated)
# 10=55 (1 mins observed)
```

## Try a few things
```{r UCparent_DirDom_propSel0pt20,eval=T}
suppressMessages(library(AlphaSimHlpR))
suppressMessages(library(tidyverse))
suppressMessages(library(genomicMateSelectR))
select <- dplyr::select

schemeDF<-read.csv(here::here("data","baselineScheme - Test.csv"), 
                   header = T, stringsAsFactors = F)

source(here::here("code","runSchemesPostBurnIn.R"))

simulations<-readRDS(here::here("output","burnIn_test.rds"))

# First the same "newBSP" used for the initial GMS test above
newBSP1<-specifyBSP(schemeDF = schemeDF,
                   nChr = 3,effPopSize = 100,quickHaplo = F,
                   segSites = 400, nQTL = 40, nSNP = 100, genVar = 40,
                   gxeVar = NULL, gxyVar = 15, gxlVar = 10,gxyxlVar = 5,
                   meanDD = 0.5,varDD = 0.01,relAA = 0.5,
                   stageToGenotype = "PYT",
                   nParents = 10, nCrosses = 4, nProgeny = 50,nClonesToNCRP = 3,
                   phenoF1toStage1 = T,errVarPreStage1 = 500,
                   useCurrentPhenoTrain = F, 
                   nCyclesToKeepRecords = 30,
                   selCritPipeAdv = selCritIID, 
                   selCritPopImprov =  selCritIID,
                   nTrainPopCycles=6,
                   nYrsAsCandidates=2,
                   maxTrainingPopSize=500,
                   modelType="DirDom",
                   propSel=0.2, # lower the effective weight on predict SD
                   crossSelCrit="UCparent")
start<-proc.time()[3]
postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
                                        newBSP=newBSP1,
                                        nPostBurnInCycles=10,
                                        selCritPop="genomicMateSelCrit",
                                        selCritPipe="selCritIID",
                                        productFunc="productPipeline",
                                        popImprovFunc="popImprovByMateSel",
                                        ncores=4,
                                        nBLASthreads=3)
saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_UCparent_DirDom_propSel0pt20.rds"))
end<-proc.time()[3]; timebsp<-end-start;
print(paste0((timebsp)/60," mins elapsed"))
# Error: Problem with `mutate()` column `SimOutput`.
# ℹ `SimOutput = future_map2(...)`.
# ✖ invalid class “Pop” object: nInd!=length(fixEff)
# Run `rlang::last_error()` to see where the error occurred.
# > rlang::last_trace()
# <error/dplyr:::mutate_error>
# Problem with `mutate()` column `SimOutput`.
# ℹ `SimOutput = future_map2(...)`.
# ✖ invalid class “Pop” object: nInd!=length(fixEff)
# Backtrace:
#      █
#   1. ├─global::runSchemesPostBurnIn(...)
#   2. │ └─`%>%`(...) code/runSchemesPostBurnIn.R:37:2
```
cbsumm13 - 24 cores
```{r UCparent_A_propSel0pt20,eval=T}
suppressMessages(library(AlphaSimHlpR))
suppressMessages(library(tidyverse))
suppressMessages(library(genomicMateSelectR))
select <- dplyr::select

schemeDF<-read.csv(here::here("data","baselineScheme - Test.csv"), 
                   header = T, stringsAsFactors = F)

source(here::here("code","runSchemesPostBurnIn.R"))

simulations<-readRDS(here::here("output","burnIn_test.rds"))

# First the same "newBSP" used for the initial GMS test above
newBSP2<-specifyBSP(schemeDF = schemeDF,
                   nChr = 3,effPopSize = 100,quickHaplo = F,
                   segSites = 400, nQTL = 40, nSNP = 100, genVar = 40,
                   gxeVar = NULL, gxyVar = 15, gxlVar = 10,gxyxlVar = 5,
                   meanDD = 0.5,varDD = 0.01,relAA = 0.5,
                   stageToGenotype = "PYT",
                   nParents = 10, nCrosses = 4, nProgeny = 50,nClonesToNCRP = 3,
                   phenoF1toStage1 = T,errVarPreStage1 = 500,
                   useCurrentPhenoTrain = F, 
                   nCyclesToKeepRecords = 30,
                   selCritPipeAdv = selCritIID, 
                   selCritPopImprov =  selCritIID,
                   nTrainPopCycles=6,
                   nYrsAsCandidates=2,
                   maxTrainingPopSize=500,
                   modelType="A", # standard additive only model
                   propSel=0.2, # lower the effective weight on predict SD
                   crossSelCrit="UCparent")
start<-proc.time()[3]
postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
                                        newBSP=newBSP2,
                                        nPostBurnInCycles=10,
                                        selCritPop="genomicMateSelCrit",
                                        selCritPipe="selCritIID",
                                        productFunc="productPipeline",
                                        popImprovFunc="popImprovByMateSel",
                                        ncores=4,
                                        nBLASthreads=5)
saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_UCparent_A_propSel0pt20.rds"))
end<-proc.time()[3]; timebsp<-end-start;
print(paste0((timebsp)/60," mins elapsed"))
# Error: Problem with `mutate()` column `SimOutput`.
# ℹ `SimOutput = future_map2(...)`.
# ✖ invalid class “Pop” object: nInd!=length(fixEff)
# 
# 10: Problem with `mutate()` column `SimOutput`.
# ℹ `SimOutput = future_map2(...)`.
# ℹ ‘MASS’ namespace cannot be unloaded:
#   namespace ‘MASS’ is imported by ‘lme4’ so cannot be unloaded
```
cbsulm05 - 64 cores
```{r MeanBV_A_propSel0pt20,eval=T}
suppressMessages(library(AlphaSimHlpR))
suppressMessages(library(tidyverse))
suppressMessages(library(genomicMateSelectR))
select <- dplyr::select

schemeDF<-read.csv(here::here("data","baselineScheme - Test.csv"), 
                   header = T, stringsAsFactors = F)

source(here::here("code","runSchemesPostBurnIn.R"))

simulations<-readRDS(here::here("output","burnIn_test.rds"))

# First the same "newBSP" used for the initial GMS test above
newBSP3<-specifyBSP(schemeDF = schemeDF,
                   nChr = 3,effPopSize = 100,quickHaplo = F,
                   segSites = 400, nQTL = 40, nSNP = 100, genVar = 40,
                   gxeVar = NULL, gxyVar = 15, gxlVar = 10,gxyxlVar = 5,
                   meanDD = 0.5,varDD = 0.01,relAA = 0.5,
                   stageToGenotype = "PYT",
                   nParents = 10, nCrosses = 4, nProgeny = 50,nClonesToNCRP = 3,
                   phenoF1toStage1 = T,errVarPreStage1 = 500,
                   useCurrentPhenoTrain = F, 
                   nCyclesToKeepRecords = 30,
                   selCritPipeAdv = selCritIID, 
                   selCritPopImprov =  selCritIID,
                   nTrainPopCycles=6,
                   nYrsAsCandidates=2,
                   maxTrainingPopSize=500,
                   modelType="A", # standard additive only model
                   propSel=0.2, # lower the effective weight on predict SD
                   crossSelCrit="MeanBV") # select based on mean BV instead of Usefulness
start<-proc.time()[3]
postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
                                        newBSP=newBSP3,
                                        nPostBurnInCycles=10,
                                        selCritPop="genomicMateSelCrit",
                                        selCritPipe="selCritIID",
                                        productFunc="productPipeline",
                                        popImprovFunc="popImprovByMateSel",
                                        ncores=4,
                                        nBLASthreads=5)
saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_MeanBV_A_propSel0pt20.rds"))
end<-proc.time()[3]; timebsp<-end-start;
print(paste0((timebsp)/60," mins elapsed"))
# Error: Problem with `mutate()` column `SimOutput`.
# ℹ `SimOutput = future_map2(...)`.
# ✖ invalid class “Pop” object: nInd!=length(fixEff)
# Run `rlang::last_error()` to see where the error occurred.
```

```{r continue after burn-in with GMS, eval=F}
# start<-proc.time()[3]
# postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
#                                         newBSP=newBSP1,
#                                         nPostBurnInCycles=10,
#                                         selCritPop="genomicMateSelCrit",
#                                         selCritPipe="selCritIID",
#                                         productFunc="productPipeline",
#                                         popImprovFunc="popImprovByMateSel",
#                                         ncores=4,
#                                         nBLASthreads=3)
# saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_newBSP1.rds"))
# end<-proc.time()[3]; timebsp1<-end-start;
# start<-proc.time()[3]
# postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
#                                         newBSP=newBSP2,
#                                         nPostBurnInCycles=10,
#                                         selCritPop="genomicMateSelCrit",
#                                         selCritPipe="selCritIID",
#                                         productFunc="productPipeline",
#                                         popImprovFunc="popImprovByMateSel",
#                                         ncores=4,
#                                         nBLASthreads=3)
# saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_newBSP2.rds"))
# end<-proc.time()[3]; timebsp2<-end-start;
# start<-proc.time()[3]
# postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
#                                         newBSP=newBSP3,
#                                         nPostBurnInCycles=10,
#                                         selCritPop="genomicMateSelCrit",
#                                         selCritPipe="selCritIID",
#                                         productFunc="productPipeline",
#                                         popImprovFunc="popImprovByMateSel",
#                                         ncores=4,
#                                         nBLASthreads=3)
# saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_newBSP3.rds"))
# end<-proc.time()[3]; timebsp3<-end-start;
```

```{r MeanBV_A_propSel0pt20,eval=T}
suppressMessages(library(AlphaSimHlpR))
suppressMessages(library(tidyverse))
suppressMessages(library(genomicMateSelectR))
select <- dplyr::select

schemeDF<-read.csv(here::here("data","baselineScheme - Test.csv"), 
                   header = T, stringsAsFactors = F)

source(here::here("code","runSchemesPostBurnIn.R"))

simulations<-readRDS(here::here("output","burnIn_test.rds"))

# First the same "newBSP" used for the initial GMS test above
newBSP4<-specifyBSP(schemeDF = schemeDF,
                   nChr = 3,effPopSize = 100,quickHaplo = F,
                   segSites = 400, nQTL = 40, nSNP = 100, genVar = 40,
                   gxeVar = NULL, gxyVar = 15, gxlVar = 10,gxyxlVar = 5,
                   meanDD = 0.5,varDD = 0.01,relAA = 0.5,
                   stageToGenotype = "PYT",
                   nParents = 10, nCrosses = 4, nProgeny = 50,nClonesToNCRP = 3,
                   phenoF1toStage1 = T,errVarPreStage1 = 500,
                   useCurrentPhenoTrain = F, 
                   nCyclesToKeepRecords = 30,
                   selCritPipeAdv = selCritIID, 
                   selCritPopImprov =  selCritIID,
                   nTrainPopCycles=6,
                   nYrsAsCandidates=2,
                   maxTrainingPopSize=500)
start<-proc.time()[3]
postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
                                        newBSP=newBSP4,
                                        nPostBurnInCycles=10,
                                        selCritPop="parentSelCritGEBV",
                                        selCritPipe="selCritIID",
                                        productFunc="productPipeline",
                                        popImprovFunc="popImprovByParentSel",
                                        ncores=4,
                                        nBLASthreads=5)
saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_ParentSel_GS.rds"))
end<-proc.time()[3]; timebsp<-end-start;
print(paste0((timebsp)/60," mins elapsed"))
```

```{r continue after burn-in with GMS, eval=F}
start<-proc.time()[3]
postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
                                        newBSP=newBSP3,
                                        nPostBurnInCycles=10,
                                        selCritPop="parentSelCritGEBV",
                                        selCritPipe="selCritIID",
                                        productFunc="productPipeline",
                                        popImprovFunc="popImprovByParentSel",
                                        ncores=4,
                                        nBLASthreads=3)
saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_newBSP3_GS.rds"))
end<-proc.time()[3]; timebsp3_gs<-end-start;
start<-proc.time()[3]
postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
                                        newBSP=newBSP3,
                                        nPostBurnInCycles=10,
                                        selCritPop="parentSelCritBLUP",
                                        selCritPipe="selCritIID",
                                        productFunc="productPipeline",
                                        popImprovFunc="popImprovByParentSel",
                                        ncores=4,
                                        nBLASthreads=3)
saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_newBSP3_PS.rds"))
end<-proc.time()[3]; timebsp3_ps<-end-start;
print(paste0((timebsp1)/60," mins elapsed"))
print(paste0((timebsp2)/60," mins elapsed"))
print(paste0((timebsp3)/60," mins elapsed"))
print(paste0((timebsp3_gs)/60," mins elapsed"))
print(paste0((timebsp3_ps)/60," mins elapsed"))
# > print(paste0((timebsp1)/60," mins elapsed"))
# [1] "61.0738166666667 mins elapsed"
# > print(paste0((timebsp2)/60," mins elapsed"))
# [1] "3.4599 mins elapsed"
# > print(paste0((timebsp3)/60," mins elapsed"))
# [1] "0.80013333333333 mins elapsed"
# > print(paste0((timebsp3_gs)/60," mins elapsed"))
# [1] "4.21411666666667 mins elapsed"
# > print(paste0((timebsp3_ps)/60," mins elapsed"))
# [1] "0.897616666666666 mins elapsed"
```
, fig.width=10
```{r plot additional test sims, fig.width=8}
forSimPlot<-readRDS(here::here("output","postBurnInGMS_test_DirDom_UCparent.rds")) %>% 
  mutate(PostBurnIn="GMS_UCparent_DirDom_propSel0pt05") %>% 
  bind_rows(readRDS(here::here("output","postBurnInGMS_test_newBSP1.rds")) %>% 
              mutate(PostBurnIn="GMS_UCparent_DirDom_propSel0pt20")) %>% 
  bind_rows(readRDS(here::here("output","postBurnInGMS_test_newBSP2.rds")) %>% 
              mutate(PostBurnIn="GMS_UCparent_A_propSel0pt20")) %>% 
  bind_rows(readRDS(here::here("output","postBurnInGMS_test_newBSP3.rds")) %>% 
              mutate(PostBurnIn="GMS_MeanBV_A_propSel0pt20")) %>% 
  bind_rows(readRDS(here::here("output","postBurnInGMS_test_newBSP3_GS.rds")) %>% 
              mutate(PostBurnIn="GS_GEBV_A")) %>% 
  bind_rows(readRDS(here::here("output","postBurnInGMS_test_newBSP3_PS.rds")) %>% 
              mutate(PostBurnIn="PS_BLUP")) %>% 
  unnest_wider(SimOutput) %>% 
  select(SimRep,PostBurnIn,records) %>% 
  unnest_wider(records) %>% 
  select(SimRep,PostBurnIn,stageOutputs) %>% 
  unnest() %>% 
  filter(stage %in% c("F1","UYT")) %>% 
  mutate(YearPostBurnIn=year-10)
#forSimPlot %>% count(PostBurnIn)
library(patchwork)
meanGplot<-forSimPlot %>% 
  filter(stage=="F1") %>% 
  group_by(PostBurnIn,YearPostBurnIn,year,.drop = F) %>% 
  summarize(meanGenMean=mean(genValMean),
            seGenMean=sd(genValMean)/n()) %>% ungroup() %>% 
  ggplot(.,aes(x=YearPostBurnIn)) +
  geom_ribbon(aes(ymin = meanGenMean - seGenMean, 
                  ymax = meanGenMean + seGenMean,
                  fill = PostBurnIn), 
              alpha=0.4) + #theme(legend.position = 'bottom') + 
  geom_line(aes(y = meanGenMean, color=PostBurnIn)) + 
  geom_vline(xintercept = 0, color='darkred')
sdGplot<-forSimPlot %>% 
  filter(stage=="F1") %>% 
  group_by(PostBurnIn,YearPostBurnIn,year,.drop = F) %>% 
  summarize(meanGenSD=mean(genValSD),
            seGenSD=sd(genValSD)/n()) %>% ungroup() %>% 
  ggplot(.,aes(x=YearPostBurnIn)) +
  geom_ribbon(aes(ymin = meanGenSD - seGenSD, 
                  ymax = meanGenSD + seGenSD,
                  fill = PostBurnIn), 
              alpha=0.4) + #theme(legend.position = 'bottom') + 
  geom_line(aes(y = meanGenSD, color=PostBurnIn)) + 
  geom_vline(xintercept = 0, color='darkred')
(meanGplot | sdGplot) + 
  plot_layout(guides = 'collect') & 
  theme_bw() & 
  theme(legend.position = 'bottom')
   
```
```{r plot additional test sims, fig.width=8}
library(patchwork)
meanGplot<-forSimPlot %>% 
  filter(stage=="F1") %>% 
  group_by(PostBurnIn,YearPostBurnIn,year,.drop = F) %>% 
  summarize(meanGenMean=mean(genValMean),
            seGenMean=sd(genValMean)/n()) %>% ungroup() %>% 
  ggplot(.,aes(x=YearPostBurnIn)) +
  geom_ribbon(aes(ymin = meanGenMean - seGenMean, 
                  ymax = meanGenMean + seGenMean,
                  fill = PostBurnIn), 
              alpha=0.4) + 
  geom_line(aes(y = meanGenMean, color=PostBurnIn)) + 
  geom_vline(xintercept = 0, color='darkred') +
  facet_grid(.~PostBurnIn,scales='free')
sdGplot<-forSimPlot %>% 
  filter(stage=="F1") %>% 
  group_by(PostBurnIn,YearPostBurnIn,year,.drop = F) %>% 
  summarize(meanGenSD=mean(genValSD),
            seGenSD=sd(genValSD)/n()) %>% ungroup() %>% 
  ggplot(.,aes(x=YearPostBurnIn)) +
  geom_ribbon(aes(ymin = meanGenSD - seGenSD, 
                  ymax = meanGenSD + seGenSD,
                  fill = PostBurnIn), 
              alpha=0.4) + 
  geom_line(aes(y = meanGenSD, color=PostBurnIn)) + 
  geom_vline(xintercept = 0, color='darkred') + 
  facet_grid(.~PostBurnIn,scales='free')
(meanGplot / sdGplot) +
  plot_layout(guides = 'collect') & 
  theme_bw() & 
  theme(legend.position = 'bottom',
        strip.background.x = element_blank(),
        strip.text.x = element_text(size=10))
```
```{r}
forSimPlot %>% 
  select(SimRep,PostBurnIn,YearPostBurnIn,year,stage,genValMean) %>% 
  filter(YearPostBurnIn>0,stage=="F1") %>% 
  spread(PostBurnIn,genValMean)
```

## Try again

MeanBV_A
UCparent_A
UCparent_DirDom_propSel0pt20
```{r}
suppressMessages(library(AlphaSimHlpR))
suppressMessages(library(tidyverse))
suppressMessages(library(genomicMateSelectR))
select <- dplyr::select

schemeDF<-read.csv(here::here("data","baselineScheme - Test.csv"), 
                   header = T, stringsAsFactors = F)

source(here::here("code","runSchemesPostBurnIn.R"))

simulations<-readRDS(here::here("output","burnIn_test.rds"))

# First the same "newBSP" used for the initial GMS test above
newBSP<-specifyBSP(schemeDF = schemeDF,
                   nChr = 3,effPopSize = 100,quickHaplo = F,
                   segSites = 400, nQTL = 40, nSNP = 100, genVar = 40,
                   gxeVar = NULL, gxyVar = 15, gxlVar = 10,gxyxlVar = 5,
                   meanDD = 0.5,varDD = 0.01,relAA = 0.5,
                   stageToGenotype = "PYT",
                   nParents = 10, nCrosses = 4, nProgeny = 50,nClonesToNCRP = 3,
                   phenoF1toStage1 = T,errVarPreStage1 = 500,
                   useCurrentPhenoTrain = F, 
                   nCyclesToKeepRecords = 30,
                   selCritPipeAdv = selCritIID, 
                   selCritPopImprov =  selCritIID,
                   nTrainPopCycles=6,
                   nYrsAsCandidates=2,
                   maxTrainingPopSize=500,
                   modelType="DirDom",
                   propSel=0.05,
                   crossSelCrit="UCparent")

# Make a few variants
newBSP1<-newBSP; newBSP1$propSel<-0.2 # lower the effective weight on predict SD
newBSP2<-newBSP1; newBSP2$modelType<-"A" # standard additive model
newBSP3<-newBSP2; newBSP3$crossSelCrit<-"MeanBV" # select based on mean BV instead of Usefulness
```

```{r}
start<-proc.time()[3]
postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
                                        newBSP=newBSP1,
                                        nPostBurnInCycles=10,
                                        selCritPop="genomicMateSelCrit",
                                        selCritPipe="selCritIID",
                                        productFunc="productPipeline",
                                        popImprovFunc="popImprovByMateSel",
                                        ncores=4,
                                        nBLASthreads=3)
saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_newBSP1.rds"))
end<-proc.time()[3]; timebsp1<-end-start;
start<-proc.time()[3]
postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
                                        newBSP=newBSP2,
                                        nPostBurnInCycles=10,
                                        selCritPop="genomicMateSelCrit",
                                        selCritPipe="selCritIID",
                                        productFunc="productPipeline",
                                        popImprovFunc="popImprovByMateSel",
                                        ncores=4,
                                        nBLASthreads=3)
saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_newBSP2.rds"))
end<-proc.time()[3]; timebsp2<-end-start;
start<-proc.time()[3]
postBurnInGMS_test<-runSchemesPostBurnIn(simulations = simulations,
                                        newBSP=newBSP3,
                                        nPostBurnInCycles=10,
                                        selCritPop="genomicMateSelCrit",
                                        selCritPipe="selCritIID",
                                        productFunc="productPipeline",
                                        popImprovFunc="popImprovByMateSel",
                                        ncores=4,
                                        nBLASthreads=3)
saveRDS(postBurnInGMS_test,file = here::here("output","postBurnInGMS_test_newBSP3.rds"))
end<-proc.time()[3]; timebsp3<-end-start;
```


## Set-up a complete comparison

Change my `AlphaSimHlpR` functions again: 

- Run a comparison of PS, GS and GMS that uses the same strategy for determining the set-of candidates
    * With vs. Without checks as candidates
    * With vs. Without the extra TP members used to make the prediction as selection candidates.
    * modelTypes: A vs. AD vs. DirDom
    * propSel: 0.01, 0.05, 0.2
    * crossSelCrit: "MeanBG", "MeanTGV", "UCparent", "UCvariety"
- Run it even longer?
- 
    * (1) consider only the most re 

